File: src/core/reasoning-extractor.ts
import { createMiddleware } from "langchain";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";
import { SystemMessage, HumanMessage } from "@langchain/core/messages";
import type { ContextGraphConfig } from "../types/config.js";
import type { ToolCall } from "../types/data-model.js";
import type { ContextualRegistry } from "./contextual-registry.js";
import { ablationFilter, filterCriticalFacts } from "./ablation-filter.js";
import { createLogger, type Logger } from "../utils/logger.js";

export function createReasoningExtractor(
  config: ContextGraphConfig,
  registry: ContextualRegistry,
  observerModel: BaseChatModel | null
) {
  const logger = createLogger(config.debug);

  return createMiddleware({
    name: "ContextGraphReasoningExtractor",

    wrapModelCall: async (request, handler) => {
      const response = await handler(request);

      // Extract content from model response
      const responseMessage = response;
      const content =
        typeof responseMessage.content === "string"
          ? responseMessage.content
          : "";

      const toolCalls = (responseMessage as any).tool_calls ?? [];

      // If no tool calls, agent is finishing — save the decision trace.
      // The trace captures the intent (user message), action (agent response),
      // and any reasoning facts from the conversation history.
      if (toolCalls.length === 0 && content) {
        const messages = request.messages ?? [];
        const facts = extractFactsFromMessages(messages);
        const capturedToolCalls = extractToolCalls(messages);
        logger.debug("Extracted %d fact(s), %d tool call(s) from %d message(s)", facts.length, capturedToolCalls.length, messages.length);
        const decision = content;

        try {
          await saveDecisionTrace(
            facts,
            decision,
            messages,
            capturedToolCalls,
            config,
            registry,
            observerModel,
            logger
          );
        } catch (err) {
          logger.warn(
            "Failed to save decision trace: %s",
            (err as Error).message
          );
        }
      }

      return response;
    },

    wrapToolCall: async (request, handler) => {
      const result = await handler(request);
      // Tool calls are captured via messages in the next model call
      return result;
    },
  });
}

/**
 * Extract reasoning facts from messages.
 * Only captures actual reasoning statements from the AI — NOT tool call
 * strings or raw tool output, which are noise and should not become constraints.
 */
function extractFactsFromMessages(messages: unknown[]): string[] {
  const facts: string[] = [];

  for (const msg of messages) {
    const message = msg as any;
    const role = message._getType?.() ?? message.role ?? "";
    const content =
      typeof message.content === "string" ? message.content : "";

    if (role === "ai" || role === "assistant") {
      // Only extract reasoning from AI messages that have NO tool calls
      // (messages with tool calls are just "I'm going to use X" — not reasoning)
      const toolCalls = message.tool_calls ?? [];
      if (toolCalls.length === 0 && content) {
        const sentences = content
          .split(/[.!?]\s+/)
          .filter((s: string) => s.trim().length > 20);
        facts.push(...sentences.slice(0, 5));
      }
    }
    // Tool results are captured separately as ToolCall nodes — they don't need
    // to be facts/constraints. The actual reasoning about tool results comes
    // from the AI's interpretation in subsequent messages.
  }

  return facts;
}

/** Extract ToolCall nodes from message history for graph storage. */
function extractToolCalls(messages: unknown[]): ToolCall[] {
  const toolCalls: ToolCall[] = [];
  const now = new Date().toISOString();

  // Build a map of tool call IDs to their results
  const toolResults = new Map<string, string>();
  for (const msg of messages) {
    const message = msg as any;
    const role = message._getType?.() ?? message.role ?? "";
    if (role === "tool") {
      const callId = message.tool_call_id ?? "";
      const content =
        typeof message.content === "string" ? message.content : "";
      if (callId) {
        toolResults.set(callId, content.substring(0, 500));
      }
    }
  }

  // Extract tool calls from AI messages
  for (const msg of messages) {
    const message = msg as any;
    const role = message._getType?.() ?? message.role ?? "";
    if (role === "ai" || role === "assistant") {
      const calls = message.tool_calls ?? [];
      for (const tc of calls) {
        toolCalls.push({
          name: tc.name,
          args: JSON.stringify(tc.args),
          result: toolResults.get(tc.id) ?? undefined,
          createdAt: now,
        });
      }
    }
  }

  return toolCalls;
}

async function saveDecisionTrace(
  facts: string[],
  decision: string,
  messages: unknown[],
  capturedToolCalls: ToolCall[],
  config: ContextGraphConfig,
  registry: ContextualRegistry,
  observerModel: BaseChatModel | null,
  logger: Logger
): Promise<void> {
  const isDiscovery = await registry.isDiscoveryMode();

  let criticalFacts = facts;
  let ablationScore: number | undefined;

  // Apply ablation filter unless in discovery mode
  if (!isDiscovery && observerModel) {
    const results = await ablationFilter(facts, decision, observerModel, logger);
    const critical = filterCriticalFacts(results);
    criticalFacts = critical.map((r) => r.fact);
    ablationScore =
      critical.length > 0
        ? critical.reduce((sum, r) => sum + r.confidence, 0) /
          critical.length
        : undefined;
  }

  // Extract intent from the first user message
  const userMessages = messages.filter(
    (m: any) => (m._getType?.() ?? m.role) === "human" || (m._getType?.() ?? m.role) === "user"
  );
  const intentDescription =
    userMessages.length > 0
      ? typeof (userMessages[0] as any).content === "string"
        ? (userMessages[0] as any).content
        : "Unknown intent"
      : "Unknown intent";

  // Use LLM for structured extraction if observer model is available
  let constraints: { description: string; type: "blocker" | "permission" | "pivot"; createdAt: string }[];
  let concepts: string[];
  let domain: string;

  if (observerModel) {
    const extraction = await extractStructuredContext(
      intentDescription,
      decision,
      criticalFacts,
      observerModel,
      logger
    );
    constraints = extraction.constraints.map((c) => ({
      ...c,
      createdAt: new Date().toISOString(),
    }));
    concepts = extraction.concepts;
    domain = config.domain ?? extraction.domain;
  } else {
    // Fallback to heuristic extraction — only create constraints from
    // facts that contain actual reasoning (not tool artifacts)
    constraints = criticalFacts
      .filter((fact) => fact.length > 20 && fact.length < 300)
      .slice(0, 5)  // Cap at 5 constraints max
      .map((fact) => ({
        description: fact,
        type: classifyFact(fact),
        createdAt: new Date().toISOString(),
      }));
    concepts = extractConceptsFallback(intentDescription, decision, criticalFacts);
    domain = config.domain ?? inferDomainFallback(intentDescription, decision);
  }

  const trace = {
    intent: {
      description: intentDescription,
      createdAt: new Date().toISOString(),
    },
    constraints,
    action: {
      description: decision.substring(0, 500),
      outcome: "pending" as const,
      createdAt: new Date().toISOString(),
    },
    justification: {
      description: criticalFacts.length > 0
        ? criticalFacts.join("; ")
        : buildJustificationSummary(intentDescription, capturedToolCalls),
      confidence: isDiscovery ? 0.5 : (ablationScore ?? 0.5),
      ablationScore,
    },
    toolCalls: capturedToolCalls,
    project: config.project,
    tenant: config.tenant,
    domain,
    agent: config.agent,
    concepts,
    status: "captured" as const,
  };

  await registry.recordDecision(trace);
}

// ── LLM-Powered Extraction ──────────────────────────────────────────────────

const EXTRACTION_SYSTEM_PROMPT = `You are a Context Extraction Engine for an AI agent's decision graph.
Given the agent's intent, decision, and facts, extract structured metadata.

Respond with valid JSON matching this schema:
{
  "domain": "<string: the domain this interaction belongs to, e.g. 'tech', 'legal', 'medical', 'finance', 'support', 'devops', or a more specific domain>",
  "concepts": ["<string: semantic tags/concepts, e.g. 'account-lockout', 'api-authentication', 'contract-review'>"],
  "constraints": [
    {
      "description": "<string: what the constraint is>",
      "type": "<'blocker' | 'permission' | 'pivot'>"
    }
  ],
  "entities": [
    {
      "label": "<string: PascalCase entity type discovered, e.g. 'ErrorPattern', 'APIEndpoint'>",
      "name": "<string: specific name of this entity>"
    }
  ]
}

Guidelines:
- **domain**: Be specific. "software-engineering" is better than "tech". Use existing domain names when applicable.
- **concepts**: Extract 1-5 semantic tags that would help find this decision trace later. Think about what future queries would match.
- **constraints**: Classify each critical fact:
  - "blocker": something that prevents or blocks an action
  - "permission": something that enables or allows an action
  - "pivot": a condition that changes the approach or priority
- **entities**: Identify domain-specific entities the agent discovered. Only include genuinely new concepts, not generic ones.

Be precise. Quality over quantity.`;

interface StructuredExtraction {
  domain: string;
  concepts: string[];
  constraints: { description: string; type: "blocker" | "permission" | "pivot" }[];
  entities: { label: string; name: string }[];
}

async function extractStructuredContext(
  intent: string,
  decision: string,
  facts: string[],
  model: BaseChatModel,
  logger: Logger
): Promise<StructuredExtraction> {
  const prompt = `## Agent Intent
${intent}

## Decision Made
${decision.substring(0, 500)}

## Critical Facts
${facts.map((f, i) => `[${i}] ${f}`).join("\n")}

Extract the domain, concepts, constraints, and entities. Respond with valid JSON only.`;

  try {
    const response = await model.invoke([
      new SystemMessage(EXTRACTION_SYSTEM_PROMPT),
      new HumanMessage(prompt),
    ]);

    const content =
      typeof response.content === "string"
        ? response.content
        : JSON.stringify(response.content);

    const jsonMatch = content.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      logger.warn("Structured extraction: no JSON found in response");
      return fallbackExtraction(intent, decision, facts);
    }

    const parsed = JSON.parse(jsonMatch[0]) as StructuredExtraction;

    // Validate and sanitize
    return {
      domain: typeof parsed.domain === "string" ? parsed.domain : "general",
      concepts: Array.isArray(parsed.concepts)
        ? parsed.concepts.filter((c): c is string => typeof c === "string").slice(0, 10)
        : [],
      constraints: Array.isArray(parsed.constraints)
        ? parsed.constraints
            .filter((c) => c && typeof c.description === "string")
            .map((c) => ({
              description: c.description,
              type: (["blocker", "permission", "pivot"].includes(c.type) ? c.type : "pivot") as "blocker" | "permission" | "pivot",
            }))
        : facts.map((f) => ({ description: f, type: "pivot" as const })),
      entities: Array.isArray(parsed.entities)
        ? parsed.entities.filter((e) => e && typeof e.label === "string")
        : [],
    };
  } catch (err) {
    logger.warn("Structured extraction failed: %s", (err as Error).message);
    return fallbackExtraction(intent, decision, facts);
  }
}

function fallbackExtraction(
  intent: string,
  decision: string,
  facts: string[]
): StructuredExtraction {
  return {
    domain: inferDomainFallback(intent, decision),
    concepts: extractConceptsFallback(intent, decision, facts),
    constraints: facts.map((f) => ({
      description: f,
      type: classifyFact(f),
    })),
    entities: [],
  };
}

// ── Heuristic Fallbacks (used when no observer model is available) ───────────

function classifyFact(fact: string): "blocker" | "permission" | "pivot" {
  const lower = fact.toLowerCase();
  if (
    lower.includes("cannot") ||
    lower.includes("error") ||
    lower.includes("fail") ||
    lower.includes("block") ||
    lower.includes("timeout") ||
    lower.includes("denied")
  ) {
    return "blocker";
  }
  if (
    lower.includes("allow") ||
    lower.includes("permit") ||
    lower.includes("access") ||
    lower.includes("grant") ||
    lower.includes("auth")
  ) {
    return "permission";
  }
  return "pivot";
}

function extractConceptsFallback(
  intent: string,
  decision: string,
  facts: string[]
): string[] {
  const combined = `${intent} ${decision} ${facts.join(" ")}`.toLowerCase();
  const concepts: string[] = [];

  const patterns: [RegExp, string][] = [
    [/\baccount\s*\w*\s*(lock|block|suspend)|lock\w*\s*(account|out)/i, "account-lockout"],
    [/\bcan'?t\s*(log\s*in|sign\s*in)|login\s*(fail|error|issue)/i, "account-lockout"],
    [/\bpassword\s*\w*\s*(reset|change|forgot)|forgot\w*\s*password/i, "password-reset"],
    [/\brate\s*limit/i, "rate-limiting"],
    [/\b429\b|\btoo many requests/i, "rate-limiting"],
    [/\bbilling|payment|invoice|refund/i, "billing"],
    [/\bapi\s*(key|token|auth)/i, "api-authentication"],
    [/\bescalat/i, "escalation"],
    [/\btimeout|latency|slow/i, "performance"],
    [/\bpermission|rbac|role|access control/i, "access-control"],
    [/\bdeployment|deploy|release/i, "deployment"],
    [/\bbug|defect|regression/i, "bug-fix"],
    [/\brefactor/i, "refactoring"],
    [/\bonboard/i, "onboarding"],
    [/\bmigrat/i, "migration"],
    [/\bdiagnos|symptom|treatment|patient/i, "clinical-decision"],
    [/\bcontract|clause|compliance|regulat/i, "compliance"],
    [/\bliabilit|negligence|statute/i, "legal-risk"],
    [/\bprescri|dosage|medication/i, "medication"],
  ];

  for (const [pattern, tag] of patterns) {
    if (pattern.test(combined) && !concepts.includes(tag)) {
      concepts.push(tag);
    }
  }

  return concepts;
}

/** Build a concise justification when no reasoning facts are available. */
function buildJustificationSummary(intent: string, toolCalls: ToolCall[]): string {
  const parts: string[] = [];
  if (toolCalls.length > 0) {
    const toolNames = [...new Set(toolCalls.map((tc) => tc.name))];
    parts.push(`Used ${toolNames.join(", ")}`);
  }
  const intentShort = intent.length > 80 ? intent.substring(0, 80) + "..." : intent;
  parts.push(`to address: ${intentShort}`);
  return parts.join(" ");
}

function inferDomainFallback(intent: string, decision: string): string {
  const combined = `${intent} ${decision}`.toLowerCase();

  if (/\bapi\b|\bendpoint|\bsdk\b|\brate.?limit|\b429\b/.test(combined)) return "tech";
  if (/\bbilling|\bpayment|\binvoice|\brefund|\bsubscription/.test(combined)) return "finance";
  if (/\baccount|\blogin|\bpassword|\bauth|\block/.test(combined)) return "support";
  if (/\blegal|\bcompliance|\bregulat|\bpolicy|\bcontract|\bliabilit/.test(combined)) return "legal";
  if (/\bmedical|\bpatient|\bdiagnos|\btreatment|\bprescri|\bsymptom/.test(combined)) return "medical";

  return "general";
}

File: python-sdk/src/graphmind_context_graphs/core/reasoning_extractor.py
"""Reasoning Extractor — captures decision traces from agent conversations."""

from __future__ import annotations

import json
import re
from typing import Any, Callable

from langchain.agents.middleware import AgentMiddleware, ModelRequest, ModelResponse
from langchain_core.language_models import BaseChatModel
from langchain_core.messages import SystemMessage, HumanMessage, AIMessage

from ..types.config import ContextGraphConfig
from ..types.data_model import DecisionTrace, Intent, Action, Justification, ToolCallRecord
from .contextual_registry import ContextualRegistry
from ..db.multi_tenant_store import RuntimeTenantContext
from ..utils.logger import create_logger


class ReasoningExtractorMiddleware(AgentMiddleware):
    """LangChain middleware that captures decision traces from agent conversations."""

    def __init__(self, config: ContextGraphConfig, registry: ContextualRegistry,
                 observer_model: BaseChatModel | None = None) -> None:
        super().__init__()
        self._config = config
        self._registry = registry
        self._observer_model = observer_model
        self._logger = create_logger(config.debug)

    def wrap_model_call(self, request: ModelRequest, handler: Callable) -> Any:
        response = handler(request)

        # Extract the AI message from ModelResponse.result
        ai_msg = None
        result_msgs = getattr(response, "result", None)
        if isinstance(result_msgs, list):
            for msg in result_msgs:
                if isinstance(msg, AIMessage):
                    ai_msg = msg
                    break

        if ai_msg is None:
            return response

        content = getattr(ai_msg, "content", "")
        if isinstance(content, list):
            content = " ".join(str(b) for b in content)
        tool_calls = getattr(ai_msg, "tool_calls", []) or []

        # If no tool calls, agent is finishing — save the decision trace
        if not tool_calls and content:
            messages = request.messages or []
            facts = _extract_facts(messages)
            captured_tools = _extract_tool_calls(messages)
            self._logger.debug("Extracted %d fact(s), %d tool call(s) from %d message(s)",
                               len(facts), len(captured_tools), len(messages))

            # Extract runtime tenant context from request
            runtime_context = self._extract_runtime_context(request)

            try:
                _save_decision_trace(
                    facts, content, messages, captured_tools,
                    self._config, self._registry, self._observer_model, self._logger,
                    runtime_context,
                )
            except Exception as e:
                self._logger.warning("Failed to save decision trace: %s", e)

        return response

    def _extract_runtime_context(self, request: ModelRequest) -> RuntimeTenantContext | None:
        """Extract runtime tenant context from the request."""
        ctx = getattr(request, "context", None)
        if not ctx:
            return None

        # Handle dict-like context
        if isinstance(ctx, dict):
            return RuntimeTenantContext(
                tenant=ctx.get("tenant"),
                project=ctx.get("project"),
                agent=ctx.get("agent"),
                agent_description=ctx.get("agent_description"),
            )

        # Handle object with attributes
        return RuntimeTenantContext(
            tenant=getattr(ctx, "tenant", None),
            project=getattr(ctx, "project", None),
            agent=getattr(ctx, "agent", None),
            agent_description=getattr(ctx, "agent_description", None),
        )


def _extract_facts(messages: list) -> list[str]:
    """Extract reasoning facts from AI messages (not tool artifacts)."""
    facts: list[str] = []
    for msg in messages:
        role = getattr(msg, "type", None) or getattr(msg, "role", "")
        content = getattr(msg, "content", "")
        if isinstance(content, list):
            content = " ".join(str(b) for b in content)

        if role in ("ai", "assistant"):
            tool_calls = getattr(msg, "tool_calls", []) or []
            if not tool_calls and content:
                sentences = re.split(r"[.!?]\s+", content)
                facts.extend(s.strip() for s in sentences if len(s.strip()) > 20)
                facts = facts[:5]
    return facts


def _extract_tool_calls(messages: list) -> list[ToolCallRecord]:
    """Extract tool call records from message history."""
    from datetime import datetime, timezone
    now = datetime.now(timezone.utc).isoformat()

    tool_results: dict[str, str] = {}
    for msg in messages:
        role = getattr(msg, "type", None) or getattr(msg, "role", "")
        if role == "tool":
            call_id = getattr(msg, "tool_call_id", "")
            content = getattr(msg, "content", "")
            if isinstance(content, list):
                content = " ".join(str(b) for b in content)
            if call_id:
                tool_results[call_id] = str(content)[:500]

    records = []
    for msg in messages:
        role = getattr(msg, "type", None) or getattr(msg, "role", "")
        if role in ("ai", "assistant"):
            for tc in (getattr(msg, "tool_calls", []) or []):
                records.append(ToolCallRecord(
                    name=tc.get("name", ""),
                    args=json.dumps(tc.get("args", {})),
                    result=tool_results.get(tc.get("id", "")),
                    created_at=now,
                ))
    return records


def _save_decision_trace(
    facts: list[str], decision: str, messages: list,
    tool_calls: list[ToolCallRecord], config: ContextGraphConfig,
    registry: ContextualRegistry, observer_model: BaseChatModel | None,
    logger: Any, runtime_context: RuntimeTenantContext | None = None,
) -> None:
    from datetime import datetime, timezone
    now = datetime.now(timezone.utc).isoformat()

    is_discovery = registry.is_discovery_mode(runtime_context)

    critical_facts = facts
    ablation_score = None

    # Extract intent from first user message
    intent_desc = "Unknown intent"
    for msg in messages:
        role = getattr(msg, "type", None) or getattr(msg, "role", "")
        if role in ("human", "user"):
            content = getattr(msg, "content", "")
            if isinstance(content, str) and content:
                intent_desc = content
            break

    # Extract runtime values from context or use config defaults
    runtime_tenant = runtime_context.tenant if runtime_context else None
    runtime_project = runtime_context.project if runtime_context else None
    runtime_agent = runtime_context.agent if runtime_context else None

    # Use LLM extraction if available, otherwise heuristic
    if observer_model:
        extraction = _extract_structured(intent_desc, decision, critical_facts, observer_model, logger)
        constraints = [{"description": c["description"], "type": c["type"], "createdAt": now} for c in extraction.get("constraints", [])]
        concepts = extraction.get("concepts", [])
        domain = config.domain or extraction.get("domain", "general")
    else:
        constraints = [
            {"description": f, "type": _classify_fact(f), "createdAt": now}
            for f in critical_facts if 20 < len(f) < 300
        ][:5]
        concepts = _extract_concepts_fallback(intent_desc, decision, critical_facts)
        domain = config.domain or _infer_domain_fallback(intent_desc, decision)

    # Build justification
    if critical_facts:
        justification_desc = "; ".join(critical_facts)
    else:
        tool_names = list({tc.name for tc in tool_calls})
        parts = []
        if tool_names:
            parts.append(f"Used {', '.join(tool_names)}")
        short_intent = intent_desc[:80] + "..." if len(intent_desc) > 80 else intent_desc
        parts.append(f"to address: {short_intent}")
        justification_desc = " ".join(parts)

    from ..types.data_model import Constraint as ConstraintModel
    trace = DecisionTrace(
        intent=Intent(description=intent_desc, created_at=now),
        constraints=[
            ConstraintModel(description=c["description"], type=c["type"], created_at=now)
            for c in constraints
        ],
        action=Action(description=decision[:500], outcome="pending", created_at=now),
        justification=Justification(
            description=justification_desc,
            confidence=0.5 if is_discovery else (ablation_score or 0.5),
            ablation_score=ablation_score,
        ),
        tool_calls=tool_calls,
        project=runtime_project or config.project,
        tenant=runtime_tenant or config.tenant,
        domain=domain,
        agent=runtime_agent or config.agent,
        concepts=concepts,
        status="captured",
        created_at=now,
        updated_at=now,
    )

    registry.record_decision(trace, runtime_context)


# ── LLM-Powered Extraction ──────────────────────────────────────────────────

EXTRACTION_PROMPT = """You are a Context Extraction Engine. Given the agent's intent, decision, and facts, extract:
{
  "domain": "<string: e.g. 'tech', 'legal', 'medical'>",
  "concepts": ["<semantic tags>"],
  "constraints": [{"description": "<what>", "type": "<blocker|permission|pivot>"}]
}
Respond with valid JSON only."""


def _extract_structured(intent: str, decision: str, facts: list[str],
                         model: BaseChatModel, logger: Any) -> dict:
    prompt = f"## Intent\n{intent}\n\n## Decision\n{decision[:500]}\n\n## Facts\n" + "\n".join(f"[{i}] {f}" for i, f in enumerate(facts))
    try:
        response = model.invoke([SystemMessage(content=EXTRACTION_PROMPT), HumanMessage(content=prompt)])
        content = str(getattr(response, "content", ""))
        match = re.search(r"\{[\s\S]*\}", content)
        if match:
            return json.loads(match.group(0))
    except Exception as e:
        logger.warning("Structured extraction failed: %s", e)
    return {"domain": "general", "concepts": [], "constraints": []}


# ── Heuristic Fallbacks ──────────────────────────────────────────────────────

def _classify_fact(fact: str) -> str:
    lower = fact.lower()
    if any(w in lower for w in ("cannot", "error", "fail", "block", "timeout", "denied")):
        return "blocker"
    if any(w in lower for w in ("allow", "permit", "access", "grant", "auth")):
        return "permission"
    return "pivot"


def _extract_concepts_fallback(intent: str, decision: str, facts: list[str]) -> list[str]:
    combined = f"{intent} {decision} {' '.join(facts)}".lower()
    patterns = [
        (r"account\s*\w*\s*(lock|block|suspend)|lock\w*\s*(account|out)", "account-lockout"),
        (r"password\s*\w*\s*(reset|change|forgot)", "password-reset"),
        (r"rate\s*limit|429|too many requests", "rate-limiting"),
        (r"billing|payment|invoice|refund", "billing"),
        (r"api\s*(key|token|auth)", "api-authentication"),
        (r"timeout|latency|slow", "performance"),
        (r"deployment|deploy|release", "deployment"),
        (r"bug|defect|regression", "bug-fix"),
        (r"contract|clause|compliance|regulat", "compliance"),
    ]
    concepts = []
    for pattern, tag in patterns:
        if re.search(pattern, combined, re.IGNORECASE) and tag not in concepts:
            concepts.append(tag)
    return concepts


def _infer_domain_fallback(intent: str, decision: str) -> str:
    combined = f"{intent} {decision}".lower()
    if re.search(r"\bapi\b|\bendpoint|\bsdk\b|\brate.?limit|\b429\b", combined): return "tech"
    if re.search(r"\bbilling|\bpayment|\binvoice|\brefund", combined): return "finance"
    if re.search(r"\baccount|\blogin|\bpassword|\bauth|\block", combined): return "support"
    if re.search(r"\blegal|\bcompliance|\bregulat|\bcontract", combined): return "legal"
    if re.search(r"\bmedical|\bpatient|\bdiagnos|\btreatment", combined): return "medical"
    return "general"
